File: app/src/main/java/com/blacklenspub/starwarsapp/FilmAdapter.java
package com.blacklenspub.starwarsapp;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

public class FilmAdapter extends RecyclerView.Adapter<FilmAdapter.FilmHolder> {

    private List<Film> films;
    private OnFilmClickListener onFilmClickListener;

    public FilmAdapter(List<Film> films, OnFilmClickListener onFilmClickListener) {
        this.films = films;
        this.onFilmClickListener = onFilmClickListener;
    }

    public void setFilms(List<Film> films) {
        this.films = films;
        notifyDataSetChanged();
    }

    @Override
    public FilmHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_film, parent, false);
        return new FilmHolder(view, onFilmClickListener);
    }

    @Override
    public void onBindViewHolder(FilmHolder holder, int position) {
        Film item = films.get(position);
        holder.film = item;
        holder.title.setText(item.getTitle());
    }

    @Override
    public int getItemCount() {
        if (films == null) {
            return 0;
        } else {
            return films.size();
        }
    }

    static class FilmHolder extends RecyclerView.ViewHolder {

        Film film;
        TextView title;

        FilmHolder(View itemView, final OnFilmClickListener onFilmClickListener) {
            super(itemView);
            title = (TextView) itemView.findViewById(R.id.tvTitle);
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    onFilmClickListener.onFilmClick(film);
                }
            });
        }
    }

    interface OnFilmClickListener {
        void onFilmClick(Film film);
    }
}

File: app/src/main/java/com/blacklenspub/starwarsapp/FilmAdapter.kt
package com.blacklenspub.starwarsapp

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import kotlinx.android.synthetic.main.item_film.view.*

class FilmAdapter(var films: List<Film>, val onFilmClickListener: OnFilmClickListener) : RecyclerView.Adapter<FilmAdapter.FilmHolder>() {

    override fun getItemCount() = films.size

    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): FilmAdapter.FilmHolder {
        val view = LayoutInflater.from(parent?.context).inflate(R.layout.item_film, parent, false)
        return FilmHolder(view, onFilmClickListener)
    }

    override fun onBindViewHolder(holder: FilmAdapter.FilmHolder, position: Int) {
        val item = films[position]
        holder.film = item
        holder.title.text = item.title
    }

    interface OnFilmClickListener {

        fun onFilmClicked(film: Film)
    }

    class FilmHolder(itemView: View, onFilmClickListener: OnFilmClickListener) : RecyclerView.ViewHolder(itemView) {

        lateinit var film: Film

        val title: TextView = itemView.tvTitle

        init {
            itemView.setOnClickListener {
                onFilmClickListener.onFilmClicked(film)
            }
        }
    }
}
